### PR TITLE
refactor: Implementation for the new mapping

### LINKED ISSUE #1854: Copy replacing doesn't work sometimes
## Description ##
Copy replacing does work as expected.

## Steps to reproduce ##
1. Open VSC
2. Try below COBOL code.
 
```cobol
       IDENTIFICATION DIVISION.                                         
       PROGRAM-ID.                                                      
           SM206A.                                                      
       DATA DIVISION.                                                   
       WORKING-STORAGE SECTION.                                         
       PROCEDURE DIVISION.                                              
       SECT-SM206-0002 SECTION.                                         
       PST-TEST-002.                                                    
           COPY                                                    KP002
                   REPLACING == WRK-DS-09V00-901                        
                                SUBTRACT 1 FROM                         
                                WRK-DS-05V00-O005-001 IN GRP-002 (1)==  
                   BY         WRK-DS-05V00-O005-001 IN WRK-XN-00050-O005
      -                  F-001 IN GRP-006 IN GRP-004 IN GRP-002 IN GRP-0
      -    01 (1).                                                      

``` 
copybook KP002.cpy
```cobol
      *HEADER,CLBRY,KP002                                                       
000100     MOVE   +00009 TO WRK-DS-05V00-O005-001  IN WRK-XN-00050-O005FKP0024.2
000200-            -001 OF GRP-006 OF GRP-004 IN GRP-003 ( 2 ).         KP0024.2
000300     ADD                                                          KP0024.2
000400         +00001 TO                                                KP0024.2
000500                   WRK-DS-09V00-901                               KP0024.2
000600                                   SUBTRACT                       KP0024.2
000700                                            1                     KP0024.2
000800                                             FROM                 KP0024.2
000900                  WRK-DS-05V00-O005-001 IN GRP-002 (1).           KP0024.2
      *END-OF,KP002                                                             
``` 
3. Analysis finishes with error. So no outline is produced.
<details><summary>LOG</summary>
<p>

	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.lambda$areaBWarning$49(CobolVisitor.java:883)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.areaBWarning(CobolVisitor.java:881)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.areaBWarning(CobolVisitor.java:876)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitStatement(CobolVisitor.java:415)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitStatement(CobolVisitor.java:76)
	at org.eclipse.lsp.cobol.core.CobolParser$StatementContext.accept(CobolParser.java:23171)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visitChildren(AbstractParseTreeVisitor.java:46)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitChildren(CobolVisitor.java:773)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.addTreeNode(CobolVisitor.java:910)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitSentence(CobolVisitor.java:463)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitSentence(CobolVisitor.java:76)
	at org.eclipse.lsp.cobol.core.CobolParser$SentenceContext.accept(CobolParser.java:22818)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visitChildren(AbstractParseTreeVisitor.java:46)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitChildren(CobolVisitor.java:773)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.addTreeNode(CobolVisitor.java:910)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.lambda$visitParagraph$15(CobolVisitor.java:250)
	at java.base/java.util.Optional.map(Optional.java:265)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitParagraph(CobolVisitor.java:248)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitParagraph(CobolVisitor.java:76)
	at org.eclipse.lsp.cobol.core.CobolParser$ParagraphContext.accept(CobolParser.java:22732)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visitChildren(AbstractParseTreeVisitor.java:46)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitChildren(CobolVisitor.java:773)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.addTreeNode(CobolVisitor.java:910)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitParagraphs(CobolVisitor.java:724)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitParagraphs(CobolVisitor.java:76)
	at org.eclipse.lsp.cobol.core.CobolParser$ParagraphsContext.accept(CobolParser.java:22647)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visitChildren(AbstractParseTreeVisitor.java:46)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitChildren(CobolVisitor.java:773)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.addTreeNode(CobolVisitor.java:910)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.lambda$visitProcedureSection$12(CobolVisitor.java:235)
	at java.base/java.util.Optional.map(Optional.java:265)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitProcedureSection(CobolVisitor.java:233)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitProcedureSection(CobolVisitor.java:76)
	at org.eclipse.lsp.cobol.core.CobolParser$ProcedureSectionContext.accept(CobolParser.java:22590)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visitChildren(AbstractParseTreeVisitor.java:46)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitChildren(CobolVisitor.java:773)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.addTreeNode(CobolVisitor.java:910)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitProcedureDivisionBody(CobolVisitor.java:729)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitProcedureDivisionBody(CobolVisitor.java:76)
	at org.eclipse.lsp.cobol.core.CobolParser$ProcedureDivisionBodyContext.accept(CobolParser.java:22527)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visitChildren(AbstractParseTreeVisitor.java:46)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitChildren(CobolVisitor.java:773)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.addTreeNode(CobolVisitor.java:910)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitProcedureDivision(CobolVisitor.java:176)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitProcedureDivision(CobolVisitor.java:76)
	at org.eclipse.lsp.cobol.core.CobolParser$ProcedureDivisionContext.accept(CobolParser.java:22025)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visitChildren(AbstractParseTreeVisitor.java:46)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitChildren(CobolVisitor.java:773)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.addTreeNode(CobolVisitor.java:910)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitProgramUnit(CobolVisitor.java:224)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitProgramUnit(CobolVisitor.java:76)
	at org.eclipse.lsp.cobol.core.CobolParser$ProgramUnitContext.accept(CobolParser.java:881)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visitChildren(AbstractParseTreeVisitor.java:46)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitChildren(CobolVisitor.java:773)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitChildren(CobolVisitor.java:76)
	at org.eclipse.lsp.cobol.core.CobolParserBaseVisitor.visitCompilationUnit(CobolParserBaseVisitor.java:27)
	at org.eclipse.lsp.cobol.core.CobolParser$CompilationUnitContext.accept(CobolParser.java:800)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visitChildren(AbstractParseTreeVisitor.java:46)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitChildren(CobolVisitor.java:773)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.lambda$visitStartRule$0(CobolVisitor.java:116)
	at java.base/java.util.Optional.map(Optional.java:265)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitStartRule(CobolVisitor.java:114)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.visitStartRule(CobolVisitor.java:76)
	at org.eclipse.lsp.cobol.core.CobolParser$StartRuleContext.accept(CobolParser.java:751)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.eclipse.lsp.cobol.core.engine.CobolLanguageEngine.transformAST(CobolLanguageEngine.java:250)
	at org.eclipse.lsp.cobol.core.engine.CobolLanguageEngine.lambda$run$3(CobolLanguageEngine.java:177)
	at org.eclipse.lsp.cobol.core.engine.analysis.Timing.measure(Timing.java:36)
	at org.eclipse.lsp.cobol.core.engine.analysis.AnalysisContext.measure(AnalysisContext.java:47)
	at org.eclipse.lsp.cobol.core.engine.CobolLanguageEngine.run(CobolLanguageEngine.java:176)
	at org.eclipse.lsp.cobol.service.delegates.validations.CobolLanguageEngineFacade.analyze(CobolLanguageEngineFacade.java:83)
	at org.eclipse.lsp.cobol.service.CobolTextDocumentService.doAnalysis(CobolTextDocumentService.java:508)
	at org.eclipse.lsp.cobol.service.CobolTextDocumentService.lambda$analyzeDocumentFirstTime$21(CobolTextDocumentService.java:486)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
11:06:20.729 [pool-1-thread-3] ERROR org.eclipse.lsp.cobol.service.CobolTextDocumentService - An exception thrown while applying analysis for file:///c%3A/Users/ap891843/IdeaProjects/slick_bkp/slick-oil/cobpgm/sm2064.CBL:
java.lang.IllegalStateException: null
	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
	at org.eclipse.lsp.cobol.common.mapping.DocumentMap.mapLocation(DocumentMap.java:143)
	at org.eclipse.lsp.cobol.common.mapping.DocumentMap.mapLocation(DocumentMap.java:132)
	at org.eclipse.lsp.cobol.common.mapping.ExtendedSource.mapLocationUnsafe(ExtendedSource.java:114)
	at org.eclipse.lsp.cobol.core.visitor.CobolVisitor.getLocality(CobolVisitor.java:818)
11:06:22.181 [pool-3-thread-1] INFO  org.eclipse.lsp.cobol.service.CobolTextDocumentService - Document closing invoked on URI file:///c%3A/Users/ap891843/IdeaProjects/slick_bkp/slick-oil/cobpgm/sm2064.CBL

</p>
</details> 

## Expected behavior ##
Code should have been analysed without any issue.

### PR DESCRIPTION
**Pull request by Nurkambay**
This refactoring will fix the following issue: #1854 

1. Pay attention to the `org.eclipse.lsp.cobol.common.mapping` package.
It contains a new implementation for the mapping. Other changes are just changes due to interface changes.

1. Also there are some changes with tests. Some tests were updated due to better mapping and a few tests were disabled because of limitations in the use cases test suit. `server/engine/src/test/java/org/eclipse/lsp/cobol/usecases`

### CODE REVIEW

**Comment by VitGottwald** [comment]
I download the artefacts (vsix-cobol-language-support-darwin-x64.zip, vsix-idms-dialect.zip, vsix-daco-dialect.zip) and tested on internal repo. It fails on every DaCo program. Will DM you on slack.

**Review by VitGottwald — CHANGES_REQUESTED** [review]
Since these changes affect the core functionality and some tests were disabled, please run the internal test suite and do a manual smoke test on the internal repo before proceeding.

**Comment by Nurkambay** [comment]
> Since these changes affect the core functionality and some tests were disabled, please run the internal test suite and do a manual smoke test on the internal repo before proceeding.

Thanks, I have added a use case test for it. Should be covered by unit tests now

**Comment by VitGottwald** [comment]
Now both IDMS and DaCo code gets properly analysed. But now I am getting a random dance of errors showing up and disappearing when enter code. Maybe worth running the _typing test_?

**Comment by Nurkambay** [comment]
> 

Looks like it is something broken in the processing itself, i will check it in dev

**Review by ishche — APPROVED** [review]

File: server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextLine.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    clear(0, size() - 1);
+  }
+
+  /**
+   * Creates a shadow copy of the line object
+   * @return a line object
+   */
+  ExtendedTextLine shadowCopy() {
+    ExtendedTextLine result = new ExtendedTextLine();
+    result.characters.clear();
```
Comment (ap891843) [inline server/common/src/main/java/org/eclipse/lsp/cobol/common/mapping/ExtendedTextLine.java:157]: I guess this is not necessary.

**Comment by Nurkambay** [comment]
> Now both IDMS and DaCo code gets properly analysed. But now I am getting a random dance of errors showing up and disappearing when enter code. Maybe worth running the _typing test_?

I have checked it. It is completely the same functionality as we have on development. Looks like when we have errors we have faster processing then expected and that's why we can see these errors. And we have the same in the development branch.

**Review by amanPras — APPROVED** [review]
👍

**Pull request merged**